Fix last-visible and last-focused

lrustand opened the pull request:
The current implementation of last-visible and last-focused is very simplified and does not correctly handle a lot of different scenarios.

The flaws in the current implementation:
- Does not update last-visible when that window is hidden.
- Does not update last-visible and last-focused when that window is closed (the variable is simply cleared).
- Does not keep track of the next visible/focused window(s) to be able to handle the above points.

I've changed the objreg keys `last-visible-main-window` and `last-focused-main-window` so that they now keep track of the focus and visibility history, instead of just pointing to one window. To reflect that they have now changed I decided to rename them to `window-visibility-history` and `window-focus-history`.

This implementation should now return the correct window in the cases described above.

`last-visible` will now only consider currently visible windows, and return the last visible/focused one of those. If there are no visible windows it will fall back to `last-focused`, just like the current implementation.

`last-focused` will unconditionally return the window that last received focus, not taking into account the visibility.

The-Compiler reviewed (changes requested):
Not a full review, but a couple of things I noticed:

- This seems to [break](https://github.com/qutebrowser/qutebrowser/actions/runs/12609600020/job/35143304080?pr=8445#step:10:138) `TestArgumentParser.test_help` in `tests/unit/commands/test_argparser.py`. Not sure off-hand why.
- Long term, we're trying to get rid of the `objreg` thing (#640), so I'd much rather just store those things outside of it, and use this change as an opportunity to get something more out of `objreg`. The simplest possibility would probably be just having two module globals in `objreg.py` (say, `window_visibility_history` and `window_focus_history`)
- By storing lists rather than single objects, you lose the main benefit of `objreg`, which is removing the objects when they have been destroyed on the Qt side (to avoid memory leaks). I'm not 100% sure off-hand if `hideEvent` is always guaranteed to be called when a widget gets destroyed, so maybe it would make more sense to use a list of [`weakref` objects](https://docs.python.org/3/library/weakref.html) instead? Or even better a [deque](https://docs.python.org/3/library/collections.html#collections.deque) so inserting things at the front doesn't need to shuffle the list around every time.
- Please [never ever use](https://realpython.com/the-most-diabolical-python-antipattern/) `except: pass`. It catches *everything* (including typos in variable names, the user pressing Ctrl-C, etc. etc.) and hides that, which is a great way to make bugs *really* frustrating to find (been there before with other projects that did that). Instead, catch whatever exception(s) actually happen.

lrustand commented:
> Not a full review, but a couple of things I noticed:

Thanks for the feedback!

> * This seems to [break](https://github.com/qutebrowser/qutebrowser/actions/runs/12609600020/job/35143304080?pr=8445#step:10:138) `TestArgumentParser.test_help` in `tests/unit/commands/test_argparser.py`. Not sure off-hand why.

Yes, I noticed, and I'm currently trying to investigate what is happening. I don't quite see how my changes could cause that, since there shouldn't really be any way that a None value is inserted into the focus or visibility history. A simple "if window is None" would probably fix the test, but I want to find out exactly what is causing it before inserting blindly an if-test.

> * Long term, we're trying to get rid of the `objreg` thing ([Get rid of objreg? #640](https://github.com/qutebrowser/qutebrowser/issues/640)), so I'd much rather just store those things outside of it, and use this change as an opportunity to get something more out of `objreg`. The simplest possibility would probably be just having two module globals in `objreg.py` (say, `window_visibility_history` and `window_focus_history`)

Certainly, will do.

> * By storing lists rather than single objects, you lose the main benefit of `objreg`, which is removing the objects when they have been destroyed on the Qt side (to avoid memory leaks). I'm not 100% sure off-hand if `hideEvent` is always guaranteed to be called when a widget gets destroyed, so maybe it would make more sense to use a list of [`weakref` objects](https://docs.python.org/3/library/weakref.html) instead? Or even better a [deque](https://docs.python.org/3/library/collections.html#collections.deque) so inserting things at the front doesn't need to shuffle the list around every time.

`hideEvent` is just used to remove it from the list of visible windows when it is not visible anymore. Windows are also removed from both lists when they are closed in `closeEvent`. Yup, deque sounds like a good data structure to store this type of data in. 

> * Please [never ever use](https://realpython.com/the-most-diabolical-python-antipattern/) `except: pass`. It catches _everything_ (including typos in variable names, the user pressing Ctrl-C, etc. etc.) and hides that, which is a great way to make bugs _really_ frustrating to find (been there before with other projects that did that). Instead, catch whatever exception(s) actually happen.

Will do.

I will convert this to a draft PR while I iron out things, and mark it as ready for review when I'm ready.

lrustand commented:
I've addressed the things you commented on, and fixed linter and test failures, so I'm ready for another review.

Regarding your suggestion to use a deque, I ended up choosing to use a list instead, but simply append to the end of the list instead of the beginning to avoid reshuffling the list on insertion. I figured since elements will be removed from arbitrary positions in the list, not just at the end, a deque does not really add any value over an end-inserted list as far as I can tell.

The-Compiler commented:
Apologies for not thinking of this earlier, but have you considered (ab)using a [`WeakValueDictionary`](https://docs.python.org/3/library/weakref.html#weakref.WeakValueDictionary) instead (with e.g. using the `win_id` as key)?

With that, values which vanished automatically get pruned by Python. Dictionaries are ordered, and a `next(reversed(d.values()))` should get you the last inserted element (or [`.popitem()`](https://docs.python.org/3/library/stdtypes.html#dict.popitem), but that also removes it unfortunately...).

lrustand commented:
Yes, I did think of that after reading the documentation for weakref, and did make an initial implementation using `WeakKeyDictionary` using the windows as keys and `None` as values. I had the same thought about (mis)using the fact that dicts are ordered in modern python versions. It worked okayish, but it had the downside that reinserting an entry does not update the position of that entry, so I still had to manually remove entries in the dictionary before updating them, and thus did not really get the full benefit of dictionaries that I had hoped for.

The-Compiler commented:
Makes a lot of sense! I have a couple of time-sensitive things to get back to the upcoming week, but I'll get back to a full review for this ASAP.

The-Compiler commented on qutebrowser/utils/objreg.py:
```diff
@@ -38,6 +39,46 @@ class CommandOnlyError(Exception):
+        Prune any None values encountered and return the last still
+        existing window from the list.
+        """
+        win = self._windows[-1]()
+        while win is None:
+            self._windows.pop()
+            win = self._windows[-1]()
+        return win
```
WDYT about avoiding the duplication here, with something like (untested):

```suggestion
        while True:
            win = self._windows[-1]()
            if win is not None:
                return win
             self._windows.pop()
```

Also perhaps this should `return None` in case there are no windows found at all, instead of raising `IndexError`? No hard feelings either-way though.

lrustand replied:
I agree. The only reason I left it as an IndexError was because the code already used exception handling to handle it. But returning None is cleaner. I can reduce a tiny bit more by pulling the check for `tabbed_browser.is_shutting_down` into the `WindowAccessHistory` as well.

lrustand commented:
@The-Compiler I was wondering if you've had a chance to look at this yet? I think I have covered all of your feedback so far, and I have nothing more I want to add to this PR. So I was wondering if I could get another review, or if we could merge this?
